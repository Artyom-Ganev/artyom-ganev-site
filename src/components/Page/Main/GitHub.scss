.page-main__github {
  &Container {
    width: 100%;
    height: 10em;
  }

  &Contributions {
    padding-top: 0.8em;

    & svg {
      overflow: visible;
    }
  }
}