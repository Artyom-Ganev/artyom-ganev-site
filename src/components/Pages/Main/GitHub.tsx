import React, { Dispatch, FunctionComponent, SetStateAction, useEffect, useState } from 'react';

import { LoadingIndicator } from 'components/Navigation';

import axios, { AxiosResponse } from 'axios';

import styles from './GitHub.module.scss';

const PROXY = 'https://urlreq.appspot.com/req?method=GET&url=';
const CONTR = 'https://github.com/users/Artyom-Ganev/contributions';
const svgRef = React.createRef<HTMLDivElement>();

/**
 * Load GitHub contribution request
 * @param {Function} update function
 */
const loadData = (update: (data: string) => void): void => {
  axios
    .get(`${PROXY}${CONTR}`)
    .then(({ data }: AxiosResponse<string>) => {
      update(data);
    })
    .catch(() => {
      update('');
    });
};

/**
 * Load GitHub contribution data hook
 */
const contributionEffectWrapper = (
  contribution: string,
  setContribution: Dispatch<SetStateAction<string>>
): void | (() => void) => {
  if (contribution) {
    return;
  }

  let needUpdate = true;
  const updater = (data: string): void => {
    if (needUpdate) {
      setContribution(data);
    }
  };
  loadData(updater);

  return (): void => {
    // useEffect was cancelled
    needUpdate = false;
  };
};

/**
 * Extract Svg Element from contribution data
 * @param {String} contribution data
 * @return {HTMLDivElement} Svg Element
 */
const extractSvg = (contribution: string): void => {
  const wrapper = document.createElement('div');
  wrapper.innerHTML = contribution;
  const svg = wrapper.querySelector<HTMLDivElement>('svg.js-calendar-graph-svg');
  if (svgRef.current) {
    svgRef.current.innerHTML = svg ? svg.outerHTML : '';
  }
};

/**
 * Github contributions chart component
 */
const GitHub: FunctionComponent = () => {
  const [contribution, setContribution] = useState<string>('');
  useEffect(() => contributionEffectWrapper(contribution, setContribution), [contribution]);
  useEffect(() => extractSvg(contribution), [contribution]);
  return (
    <div className={`flexBox justifyContentCenter ${styles.container}`}>
      <div ref={svgRef} className={styles.contributions} />
      {!svgRef.current && <LoadingIndicator />}
    </div>
  );
};

export default GitHub;
