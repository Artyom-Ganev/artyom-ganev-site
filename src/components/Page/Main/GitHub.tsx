import axios, { AxiosResponse } from 'axios';
import React, { useEffect, useState } from 'react';
import LoadingIndicator from '../../Navigation/LoadingIndicator';
import './GitHub.scss';

const PROXY = 'https://urlreq.appspot.com/req?method=GET&url=';
const CONTR = 'https://github.com/users/Artyom-Ganev/contributions';
const svgRef = React.createRef<HTMLDivElement>();

/**
 * Load GitHub contribution hook
 */
const useContribution = () => {
  const [contribution, setContribution] = useState<string>('');
  useEffect(() => {
    let cancel = false;
    if (!contribution) {
      (async () => {
        const { data }: AxiosResponse<string> = await axios.get(`${PROXY}${CONTR}`);
        if (!cancel) {
          setContribution(data);
        }
      })();
    }
    return () => {
      cancel = true;
    };
  });
  return contribution;
};

/**
 * Github contributions chart component
 */
const GitHub = () => {
  const data = useContribution();
  if (svgRef.current) {
    const wrapper = document.createElement('div');
    wrapper.innerHTML = data;
    const contribution = wrapper.querySelector('svg.js-calendar-graph-svg');
    svgRef.current.innerHTML = contribution ? contribution.outerHTML : '';
  }
  return (
    <div className='flexBox justifyContentCenter page-main__githubContainer'>
      <div ref={svgRef} className='page-main__githubContributions' />
      {!svgRef.current && <LoadingIndicator />}
    </div>
  );
};

export default GitHub;
