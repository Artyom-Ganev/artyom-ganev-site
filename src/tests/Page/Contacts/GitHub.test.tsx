import React from 'react';
import renderer from 'react-test-renderer';
import GitHub from '../../../components/Page/Main/GitHub';

test('GitHub', () => {
  const component = renderer.create(<GitHub />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
