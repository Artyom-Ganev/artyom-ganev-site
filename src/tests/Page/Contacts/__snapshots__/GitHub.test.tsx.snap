// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub 1`] = `
<div
  className="flexBox justifyContentCenter container"
>
  <div
    className="contributions"
  />
  <div
    className="flexBox flexGrow-1 flexShrink-1 justifyContentCenter"
  >
    <div
      className="container"
    />
  </div>
</div>
`;
